program-entrypoint/src/lib.rs

```diff
@@ -281,18 +281,61 @@ macro_rules! custom_panic_default {
 
 /// The bump allocator used as the default rust heap when running programs.
 pub struct BumpAllocator {
+    #[deprecated(
+        since = "2.2.2",
+        note = "This field should not be accessed directly. It will become private in future versions"
+    )]
     pub start: usize,
+    #[deprecated(
+        since = "2.2.2",
+        note = "This field should not be accessed directly. It will become private in future versions"
+    )]
     pub len: usize,
 }
+
+impl BumpAllocator {
+    /// Creates the allocator tied to a provided slice.
+    /// This will not initialize the provided memory, except for the first
+    /// bytes where the pointer is stored.
+    ///
+    /// # Safety
+    /// As long as BumpAllocator or any of its allocations are alive,
+    /// writing into or deallocating the arena will cause UB.
+    ///
+    /// Integer arithmetic in this global allocator implementation is safe when
+    /// operating on the prescribed `HEAP_START_ADDRESS` and `HEAP_LENGTH`. Any
+    /// other use may overflow and is thus unsupported and at one's own risk.
+    #[inline]
+    #[allow(clippy::arithmetic_side_effects)]
+    pub unsafe fn new(arena: &mut [u8]) -> Self {
+        debug_assert!(
+            arena.len() > size_of::<usize>(),
+            "Arena should be larger than usize"
+        );
+
+        // create a pointer to the start of the arena
+        // that will hold an address of the byte following free space
+        let pos_ptr = arena.as_mut_ptr() as *mut usize;
+        // initialize the data there
+        *pos_ptr = pos_ptr as usize + arena.len();
```

Comment:
I think there is a potential alignment issue here.
`arena: &mut [u8]` is not guaranteed to have proper alignment for it to be cast to a pointer to a `usize`.
A `[u8]` can have any alignment, while a pointer to `usize` needs to have an alignment of 4 or higher.

You can use [`write_unaligned()`](https://doc.rust-lang.org/std/ptr/fn.write_unaligned.html) to avoid the undefined behavior.
[`write_unaligned()` example](https://doc.rust-lang.org/std/ptr/fn.write_unaligned.html#examples) is almost literary the code you are writing here.

Reply:
Write unaligned could work, yes, but I think to properly address this, the entire BumpAllocator should be rewritten to use some form of interior mutability in the struct itself rather than storing the position pointer inline with the allocated data. It would make code a lot cleaner too. I will make a PR for that.